third_party/move/move-compiler/src/parser/syntax.rs

```diff
@@ -1464,21 +1467,80 @@ fn parse_for_loop(context: &mut Context) -> Result<(Exp, bool), Box<Diagnostic>>
 
 // Match = "match" "(" <Exp> ")" "{" ( <MatchArm> ","? )* "}"
 // MatchArm = <Bind> ( "if" <Exp> )? "=>" <Exp>
+// If called, we know we are looking at `match (`.
 fn parse_match_exp(context: &mut Context) -> Result<Exp, Box<Diagnostic>> {
+    // We cannot uniquely determine this is actually a match expression
+    // until we have seen the `match (exp) {` prefix. We parse the parts and
```

Comment:
Ah - requiring `if` and `while` conditions have `()` around them saves us from `if match() {} else {}`.

I think it would be worth adding a test case with `fun match() {}`: it parses fine (because we are not in an expression matching context).

Reply:
Added the test

Reply:
BTW, if we would not have the `if (..)` nonsense, we would not have a problem -- we could easily identify `match e` because there is no juxtaposition in expressions ...